File: voice-relay-ts/src/oauth/provider.ts
/**
 * OAuthServerProvider — the authorization server, Google upstream.
 *
 * Ported from the security-reviewed Python provider, adapted to the TS SDK
 * interface (verified against @modelcontextprotocol/sdk@1.x):
 *   - authorize(client, params, res) writes the redirect to Google onto res;
 *   - the SDK verifies PKCE via challengeForAuthorizationCode();
 *   - verifyAccessToken() returns AuthInfo with extra.slackUid, which the
 *     tool reads.
 *
 * Safety (see docs/voice-relay-oauth.md): identity comes only from Google's
 * verified email → Slack UID, gated by allowed domain + Slack membership;
 * codes are one-time + client-bound; refresh rotates; the SDK owns PKCE +
 * redirect_uri validation. The Google callback route calls
 * completeGoogleCallback() (not part of the SDK interface).
 */

import { randomBytes } from "node:crypto";
import type { Response } from "express";
import type {
  AuthorizationParams,
  OAuthServerProvider,
} from "@modelcontextprotocol/sdk/server/auth/provider.js";
import type { OAuthRegisteredClientsStore } from "@modelcontextprotocol/sdk/server/auth/clients.js";
import type { AuthInfo } from "@modelcontextprotocol/sdk/server/auth/types.js";
import { InvalidTokenError } from "@modelcontextprotocol/sdk/server/auth/errors.js";
import type {
  OAuthClientInformationFull,
  OAuthTokenRevocationRequest,
  OAuthTokens,
} from "@modelcontextprotocol/sdk/shared/auth.js";

import { exchangeCodeForEmail, googleAuthUrl } from "./google.js";
import { OAuthStore } from "./store.js";

const ACCESS_TTL_MS = 3_600_000; // 1h
const REFRESH_TTL_MS = 30 * 24 * 3_600_000; // 30d
const CODE_TTL_MS = 300_000; // 5m

/** Domain/membership denial — the callback route maps this to 403. */
export class ForbiddenError extends Error {}

export type ExchangeEmail = (opts: {
  code: string;
  clientId: string;
  clientSecret: string;
  redirectUri: string;
}) => Promise<string>;

export interface ProviderOptions {
  publicUrl: string;
  googleClientId: string;
  googleClientSecret: string;
  allowedDomains: Set<string>;
  resolveUid: (email: string) => Promise<string>;
  store: OAuthStore;
  /** Injectable Google exchange (defaults to the real one; tests supply a fake). */
  exchangeEmail?: ExchangeEmail;
}

export class HoneybotOAuthProvider implements OAuthServerProvider {
  private readonly googleRedirect: string;
  private readonly exchangeEmail: ExchangeEmail;

  constructor(private readonly opts: ProviderOptions) {
    this.googleRedirect = `${opts.publicUrl.replace(/\/+$/, "")}/oauth/callback`;
    this.exchangeEmail = opts.exchangeEmail ?? exchangeCodeForEmail;
  }

  get clientsStore(): OAuthRegisteredClientsStore {
    const store = this.opts.store;
    return {
      getClient(clientId: string): OAuthClientInformationFull | undefined {
        return store.getClient(clientId);
      },
      registerClient(client: OAuthClientInformationFull): OAuthClientInformationFull {
        store.putClient(client);
        return client;
      },
    };
  }

  async authorize(
    client: OAuthClientInformationFull,
    params: AuthorizationParams,
    res: Response,
  ): Promise<void> {
    // The SDK already validated params.redirectUri against the client's
    // registered redirect_uris. Stash under our single-use state, go to Google.
    const state = token(24);
    this.opts.store.putPending(state, {
      clientId: client.client_id,
      redirectUri: params.redirectUri,
      codeChallenge: params.codeChallenge,
      clientState: params.state ?? "",
      scopes: params.scopes ?? [],
    });
    res.redirect(
      googleAuthUrl({
        clientId: this.opts.googleClientId,
        redirectUri: this.googleRedirect,
        state,
      }),
    );
  }

  /** Google's redirect handler (custom route). Returns the client redirect URL. */
  async completeGoogleCallback(code: string, state: string): Promise<string> {
    const pending = this.opts.store.popPending(state);
    if (!pending) throw new Error("unknown or expired authorization state");

    const email = await this.exchangeEmail({
      code,
      clientId: this.opts.googleClientId,
      clientSecret: this.opts.googleClientSecret,
      redirectUri: this.googleRedirect,
    });
    const domain = email.split("@").pop() ?? "";
    if (!this.opts.allowedDomains.has(domain)) {
      throw new ForbiddenError(`email domain not allowed: ${email}`);
    }
    const uid = await this.opts.resolveUid(email);
    if (!uid) throw new ForbiddenError(`no Slack user for ${email}`);

    const ourCode = token(24);
    this.opts.store.putCode(ourCode, {
      code: ourCode,
      clientId: pending.clientId,
      redirectUri: pending.redirectUri,
      codeChallenge: pending.codeChallenge,
      scopes: pending.scopes,
      slackUid: uid,
      email,
      expiresAt: Date.now() + CODE_TTL_MS,
    });
    return redirectWithCode(pending.redirectUri, ourCode, pending.clientState);
  }

  async challengeForAuthorizationCode(
    client: OAuthClientInformationFull,
    authorizationCode: string,
  ): Promise<string> {
    const c = this.opts.store.getCode(authorizationCode);
    if (!c || c.clientId !== client.client_id || c.expiresAt < Date.now()) {
      throw new Error("invalid authorization code");
    }
    return c.codeChallenge;
  }

  async exchangeAuthorizationCode(
    client: OAuthClientInformationFull,
    authorizationCode: string,
  ): Promise<OAuthTokens> {
    // PKCE + redirect_uri already validated by the SDK. One-time use.
    const c = this.opts.store.getCode(authorizationCode);
    if (!c || c.clientId !== client.client_id || c.expiresAt < Date.now()) {
      throw new Error("invalid authorization code");
    }
    this.opts.store.popCode(authorizationCode);
    return this.issue(client.client_id, c.scopes, c.slackUid, c.email);
  }

  async exchangeRefreshToken(
    client: OAuthClientInformationFull,
    refreshToken: string,
    scopes?: string[],
  ): Promise<OAuthTokens> {
    const rec = this.opts.store.getRefresh(refreshToken);
    if (!rec || rec.clientId !== client.client_id) {
      throw new Error("unknown refresh token");
    }
    this.opts.store.revoke(refreshToken); // rotate
    return this.issue(client.client_id, scopes?.length ? scopes : rec.scopes, rec.slackUid, rec.email);
  }

  async verifyAccessToken(accessToken: string): Promise<AuthInfo> {
    const rec = this.opts.store.getAccess(accessToken);
    if (!rec || (rec.expiresAt && rec.expiresAt < Date.now())) {
      throw new InvalidTokenError("invalid or expired token");
    }
    return {
      token: accessToken,
      clientId: rec.clientId,
      scopes: rec.scopes,
      expiresAt: Math.floor(rec.expiresAt / 1000),
      extra: { slackUid: rec.slackUid, email: rec.email },
    };
  }

  async revokeToken(
    _client: OAuthClientInformationFull,
    request: OAuthTokenRevocationRequest,
  ): Promise<void> {
    this.opts.store.revoke(request.token);
  }

  private issue(clientId: string, scopes: string[], slackUid: string, email: string): OAuthTokens {
    const access = token(32);
    const refresh = token(32);
    const now = Date.now();
    this.opts.store.putAccess(access, {
      slackUid,
      email,
      clientId,
      scopes,
      expiresAt: now + ACCESS_TTL_MS,
      refreshToken: refresh,
    });
    this.opts.store.putRefresh(refresh, {
      slackUid,
      email,
      clientId,
      scopes,
      expiresAt: now + REFRESH_TTL_MS,
    });
    return {
      access_token: access,
      token_type: "Bearer",
      expires_in: Math.floor(ACCESS_TTL_MS / 1000),
      scope: scopes.length ? scopes.join(" ") : undefined,
      refresh_token: refresh,
    };
  }
}

function token(bytes: number): string {
  return randomBytes(bytes).toString("base64url");
}

function redirectWithCode(redirectUri: string, code: string, clientState?: string): string {
  const sep = redirectUri.includes("?") ? "&" : "?";
  const q = new URLSearchParams({ code });
  if (clientState) q.set("state", clientState);
  return `${redirectUri}${sep}${q.toString()}`;
}

File: voice-relay/voice_relay/oauth/provider.py
"""OAuthAuthorizationServerProvider — the authorization server, Google upstream.

Implements the mcp SDK provider interface. The SDK owns the HTTP endpoints
(/authorize, /token, /register, metadata) and PKCE verification; this class
owns the decisions: register clients (DCR), redirect /authorize to Google,
turn Google's callback into our own one-time authorization code carrying the
resolved Slack UID, and issue/validate opaque tokens.

Security properties (for review):
  - `state` binds our Google round-trip to the client's /authorize request;
    it's single-use (pop_pending).
  - The SDK validates the client's redirect_uri (against its DCR record) and
    PKCE before our code is ever exchanged.
  - We refuse to mint a code unless Google says the email is verified, its
    domain is allow-listed, AND it resolves to a Slack UID.
  - Authorization codes are one-time (pop on exchange). Access/refresh
    tokens are opaque and server-validated; refresh rotates.
"""

from __future__ import annotations

import asyncio
import secrets
import time
import urllib.parse

from mcp.server.auth.provider import RefreshToken
from mcp.shared.auth import OAuthClientInformationFull, OAuthToken
from pydantic import AnyUrl

from voice_relay.oauth import google
from voice_relay.oauth.models import HoneybotAccessToken, HoneybotAuthorizationCode
from voice_relay.oauth.store import Store, TokenRecord

ACCESS_TTL = 3600               # 1h access tokens
REFRESH_TTL = 30 * 24 * 3600    # 30d refresh tokens
CODE_TTL = 300                  # 5m authorization codes


class HoneybotOAuthProvider:
    def __init__(
        self,
        *,
        public_url: str,
        google_client_id: str,
        google_client_secret: str,
        allowed_domains,
        slack_uid_resolver,   # callable(email) -> slack_uid | ""
        store: Store,
    ):
        self._public_url = public_url.rstrip("/")
        self._google_redirect = f"{self._public_url}/oauth/callback"
        self._gid = google_client_id
        self._gsecret = google_client_secret
        self._domains = {d.strip().lower() for d in allowed_domains if d.strip()}
        self._resolve_uid = slack_uid_resolver
        self._store = store

    # ---- DCR ---------------------------------------------------------------
    async def get_client(self, client_id: str):
        raw = self._store.get_client(client_id)
        return OAuthClientInformationFull.model_validate(raw) if raw else None

    async def register_client(self, client_info: OAuthClientInformationFull) -> None:
        self._store.put_client(client_info.client_id, client_info.model_dump(mode="json"))

    # ---- authorize -> Google ----------------------------------------------
    async def authorize(self, client: OAuthClientInformationFull, params) -> str:
        # The SDK has already validated params.redirect_uri against the
        # client's registered redirect_uris. Stash the request under our own
        # single-use state and send the user to Google.
        state = secrets.token_urlsafe(24)
        self._store.put_pending(state, {
            "client_id": client.client_id,
            "redirect_uri": str(params.redirect_uri),
            "redirect_uri_provided_explicitly": params.redirect_uri_provided_explicitly,
            "code_challenge": params.code_challenge,
            "client_state": params.state,
            "scopes": params.scopes or [],
        })
        return google.auth_url(
            client_id=self._gid, redirect_uri=self._google_redirect, state=state
        )

    async def complete_google_callback(self, code: str, state: str) -> str:
        """Handle Google's redirect. Returns the URL to send the client back to.

        Called by the /oauth/callback custom route. Raises on any failure so
        the route can render an error instead of redirecting.
        """
        pending = self._store.pop_pending(state)
        if not pending:
            raise ValueError("unknown or expired authorization state")

        # Google exchange + Slack lookup are blocking urllib — run them off
        # the event loop so a login doesn't stall the relay.
        email = await asyncio.to_thread(
            google.exchange_code_for_email,
            code=code, client_id=self._gid,
            client_secret=self._gsecret, redirect_uri=self._google_redirect,
        )
        domain = email.rsplit("@", 1)[-1]
        if not self._domains or domain not in self._domains:
            raise PermissionError(f"email domain not allowed: {email}")
        uid = await asyncio.to_thread(self._resolve_uid, email)
        if not uid:
            raise PermissionError(f"no Slack user for {email}")

        our_code = secrets.token_urlsafe(24)
        self._store.put_code(our_code, HoneybotAuthorizationCode(
            code=our_code,
            scopes=pending["scopes"],
            expires_at=time.time() + CODE_TTL,
            client_id=pending["client_id"],
            code_challenge=pending["code_challenge"],
            redirect_uri=AnyUrl(pending["redirect_uri"]),
            redirect_uri_provided_explicitly=pending["redirect_uri_provided_explicitly"],
            slack_uid=uid,
            email=email,
        ))
        return _redirect_with_code(
            pending["redirect_uri"], our_code, pending.get("client_state")
        )

    # ---- code + token ------------------------------------------------------
    async def load_authorization_code(self, client, authorization_code: str):
        code = self._store.get_code(authorization_code)
        if not code or code.client_id != client.client_id:
            return None
        if code.expires_at < time.time():
            return None
        return code

    async def exchange_authorization_code(self, client, authorization_code) -> OAuthToken:
        # PKCE + redirect_uri already checked by the SDK. One-time use.
        self._store.pop_code(authorization_code.code)
        return self._issue(
            client.client_id, authorization_code.scopes,
            authorization_code.slack_uid, authorization_code.email,
        )

    async def load_access_token(self, token: str):
        rec = self._store.get_access(token)
        if not rec or (rec.expires_at and rec.expires_at < time.time()):
            return None
        return HoneybotAccessToken(
            token=token, client_id=rec.client_id, scopes=rec.scopes,
            expires_at=rec.expires_at, slack_uid=rec.slack_uid, email=rec.email,
        )

    async def load_refresh_token(self, client, refresh_token: str):
        rec = self._store.get_refresh(refresh_token)
        if not rec or rec.client_id != client.client_id:
            return None
        return RefreshToken(
            token=refresh_token, client_id=rec.client_id,
            scopes=rec.scopes, expires_at=rec.expires_at or None,
        )

    async def exchange_refresh_token(self, client, refresh_token, scopes) -> OAuthToken:
        rec = self._store.get_refresh(refresh_token.token)
        if not rec:
            raise ValueError("unknown refresh token")
        self._store.revoke(refresh_token.token)  # rotate
        return self._issue(client.client_id, scopes or rec.scopes, rec.slack_uid, rec.email)

    async def revoke_token(self, token) -> None:
        self._store.revoke(getattr(token, "token", token))

    def _issue(self, client_id, scopes, slack_uid, email) -> OAuthToken:
        access = secrets.token_urlsafe(32)
        refresh = secrets.token_urlsafe(32)
        now = int(time.time())
        scopes = list(scopes or [])
        self._store.put_access(access, TokenRecord(
            slack_uid=slack_uid, email=email, client_id=client_id,
            scopes=scopes, expires_at=now + ACCESS_TTL, refresh_token=refresh,
        ))
        self._store.put_refresh(refresh, TokenRecord(
            slack_uid=slack_uid, email=email, client_id=client_id,
            scopes=scopes, expires_at=now + REFRESH_TTL,
        ))
        return OAuthToken(
            access_token=access, token_type="Bearer", expires_in=ACCESS_TTL,
            scope=" ".join(scopes) if scopes else None, refresh_token=refresh,
        )


def _redirect_with_code(redirect_uri: str, code: str, client_state) -> str:
    sep = "&" if "?" in redirect_uri else "?"
    q = {"code": code}
    if client_state:
        q["state"] = client_state
    return redirect_uri + sep + urllib.parse.urlencode(q)
